django/utils/http.py

```diff
@@ -303,3 +304,146 @@ def is_safe_url(url, host=None):
         return False
     return ((not url_info.netloc or url_info.netloc == host) and
             (not url_info.scheme or url_info.scheme in ['http', 'https']))
+
+
+# Functions lifted from their respective modules in order to implement
+# several checks which were not possible with the stdlib, namely limiting
+# the number of fields that will be parsed.
+#
+# Copyright (C) 2013 Python Software Foundation, see license.python.txt for
+# details.
+def limited_parse_qsl_py2(qs, keep_blank_values=0, strict_parsing=0, fields_limit=None):
+    """
+    Parse a query given as a string argument and return a list.
+
+    Lifted from urlparse with an additional "fields_limit" argument.
+
+    Arguments:
+
+    qs: percent-encoded query string to be parsed
+
+    keep_blank_values: flag indicating whether blank values in
+        percent-encoded queries should be treated as blank strings.  A
+        true value indicates that blanks should be retained as blank
+        strings.  The default false value indicates that blank values
+        are to be ignored and treated as if they were  not included.
+
+    strict_parsing: flag indicating what to do with parsing errors. If
+        false (the default), errors are silently ignored. If true,
+        errors raise a ValueError exception.
+
+    fields_limit: maximum number of fields parsed, or an exception
+        is raised. None means no limit and is the default.
+    """
+    fields_limit = fields_limit or 0
```

Comment:
I agree that `None` is a better name for "no limit" than 0. This line still looks odd. I'd rewrite 338-341 as:
```python
    if fields_limit:
        pairs = re.split('[&;]', qs, fields_limit)
        if len(pairs) > fields_limit:
             raise TooManyFieldsSent('Too many fields sent. Check DATA_UPLOAD_MAX_NUMBER_FIELDS.')
    else:
        pairs = re.split('[&;]', qs)
```
(the comment below on caching the regex notwithstanding)